### PR TITLE
[DEVEX-178] Module Locals Refactored to follow a common pattern

### PR DESCRIPTION
**Pull request by mamu0**
### List of changes

Updated naming convention module with `locals.tf` and `outputs.tf` files.
Updated module locals for:

- `azure_api_management`
- `azure_app_services`
- `azure_app_services_exposed`
- `azure_function_app`
- `azure_function_app_exposed`
- `azure_eventhub`
- `azure_postgres_server`
- `azure_storage_account`

Added examples for:

- `azure_app_services`
- `azure_app_services_exposed`
- `azure_function_app`
- `azure_function_app_exposed`

### Motivation and context

Since many modules shared common parts in their local definitions, the `naming_convention` module was updated to include `outputs` that generate the names of various resources. 
This change allows the module to be reused across different projects and ensures a standardized naming convention. Additionally, `examples` directories have been added to the modified modules to facilitate easier testing and demonstration.

### Type of changes

- [ ] Add new resources
- [x] Update configuration to existing resources
- [ ] Remove existing resources

### Does this introduce a change to production resources with possible user impact?

- [ ] Yes, users may be impacted applying this change
- [x] No

### CODE REVIEW

**Comment by christian-calabrese** [comment]
Amazing job!! Do you think we can also remove (or make them optional for retro compatibility) and populate automatically common variables as:
- subnet_pep_id
- private_dns_zone_resource_group_name
- action_group_id
- tags (to understand how to set the CostCenter)

I think this might highly improve the amount of boilerplate code needed in the modules user's repositories

**Comment by mamu0** [comment]
> Amazing job!! Do you think we can also remove (or make them optional for retro compatibility) and populate automatically common variables as:
> 
> * subnet_pep_id
> * private_dns_zone_resource_group_name
> * action_group_id
> * tags (to understand how to set the CostCenter)
> 
> I think this might highly improve the amount of boilerplate code needed in the modules user's repositories

Thank you!!
Yes, I considered that, but I didn’t make the change because the modules currently differ in how they handle these variables. For example, in API management, you might have two different names and no one pep. 
However, I agree that we should establish a standard, and I believe the `naming_convention` module is the right candidate for this.

As for the tags, it’s a bit different. Right now, they are defined as `map(any)`, and while similar tags are used, they aren’t standardized across the repo. We should identify the common tags and possibly `merge` them with additional tags passed externally.

But in general i agree with you suggestion! 😄

**Comment by gunzip** [comment]
I propose to merge this one and try the suggest refactoring in another PR

File: infra/modules/azure_function_app/locals.tf [hunk 1]
```diff
@@ -39,11 +34,11 @@ locals {
 
   storage_account = {
     replication_type = local.tier == "s" ? "LRS" : "ZRS"
-    name             = replace("${local.project}${replace(local.domain, "-", "")}${var.environment.app_name}stfn${var.environment.instance_number}", "-", "")
-    pep_blob_name    = "${local.app_name_prefix}-blob-pep-${var.environment.instance_number}"
-    pep_file_name    = "${local.app_name_prefix}-file-pep-${var.environment.instance_number}"
-    pep_queue_name   = "${local.app_name_prefix}-queue-pep-${var.environment.instance_number}"
-    alert            = "[${replace("${local.project}${replace(local.domain, "-", "")}${var.environment.app_name}stfn${var.environment.instance_number}", "-", "")}] Low Availability"
+    name             = replace("${module.naming_convention.project}${replace(module.naming_convention.domain, "-", "")}${var.environment.app_name}stfn${module.naming_convention.suffix}", "-", "")
```
Comment (gunzip) [inline infra/modules/azure_function_app/locals.tf:37]: what about lowercasing and limiting to 24 chars? something like:

```
  name = substr( replace(lower("${module.naming_convention.prefix}st${module.naming_convention.suffix}"), "/[^a-z0-9]/", ""),
    0,
    24
  )

```

File: infra/modules/azure_function_app/data.tf [hunk 2]
```diff
@@ -13,11 +13,6 @@ data "azurerm_private_dns_zone" "storage_account_file" {
   resource_group_name = local.private_dns_zone.resource_group_name
 }
 
-data "azurerm_private_dns_zone" "storage_account_table" {
```
Comment (gunzip) [inline infra/modules/azure_function_app/data.tf:16]: why remove this?

File: infra/modules/azure_function_app/examples/complete/locals.tf [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+  environment = {
+    prefix          = "io"
+    env_short       = "p"
+    location        = "italynorth"
+    domain          = "modules"
+    app_name        = "test"
+    instance_number = "01"
+  }
+
+  location_map = {
```
Comment (gunzip) [inline infra/modules/azure_function_app/examples/complete/locals.tf:11]: can't we take `location_map` from the naming convention module?

**Review by gunzip — CHANGES_REQUESTED** [review]
are these changes backward compatibles? in case we need a major version bump + steps to migrate inside PR description

File: infra/modules/azure_function_app/data.tf (hunk 2 above)
Comment (mamu0) [reply infra/modules/azure_function_app/data.tf:16]: Because it never gets called, and the pre-commit `tflint` failed

**Comment by mamu0** [comment]
> I propose to merge this one and try the suggest refactoring in another PR

Yes, I agree with you!

File: infra/modules/azure_function_app/locals.tf (hunk 1 above)
Comment (mamu0) [reply infra/modules/azure_function_app/locals.tf:37]: Regarding the 24 character limit, it is already done during the validation phase of the environment variable, and it is better to leave it like that, doing checks in locals can lead to drift in the state in case of errors during apply. 

The `lower(...)` can be done, it must first be checked if a function with capital letters has ever been generated, because otherwise backward compatibility would be lost

File: infra/modules/azure_function_app/examples/complete/locals.tf (hunk 3 above)
Comment (mamu0) [reply infra/modules/azure_function_app/examples/complete/locals.tf:11]: we can and we do it 😆

File: infra/modules/azure_api_management/example/complete/main.tf [hunk 4]
```diff
@@ -1,3 +1,16 @@
...
+  source = "../../../azure_naming_convention"
+
+  environment = {
+    prefix          = local.environment.prefix
+    env_short       = local.environment.env_short
+    location        = local.environment.location
+    domain          = local.environment.domain
+    app_name        = local.environment.app_name
+    instance_number = local.environment.instance_number
+  }
```
Comment (gunzip) [inline infra/modules/azure_api_management/example/complete/main.tf:11]: can't we just do.. 

```suggestion
  environment = local.environment
```

?

File: infra/modules/azure_event_hub/example/complete/main.tf [hunk 5]
```diff
@@ -1,3 +1,16 @@
+module "naming_convention" {
+  source = "../../../azure_naming_convention"
+
+  environment = {
```
Comment (gunzip) [inline infra/modules/azure_event_hub/example/complete/main.tf:4]: ditto

File: infra/modules/azure_function_app/examples/complete/main.tf [hunk 6]
```diff
@@ -0,0 +1,51 @@
+module "naming_convention" {
+  source = "../../../azure_naming_convention"
+
+  environment = {
```
Comment (gunzip) [inline infra/modules/azure_function_app/examples/complete/main.tf:4]: same here

**Comment by gunzip** [comment]
lgtm, let's see what @pagopa/engineering-team-devex think

File: infra/modules/azure_api_management/locals.tf [hunk 7]
```diff
@@ -1,18 +1,7 @@
...
-  location_short   = lookup(local.location_map, var.environment.location, "neu")
-  project          = "${var.environment.prefix}-${var.environment.env_short}-${local.location_short}"
-  domain           = var.environment.domain == null ? "-" : "-${var.environment.domain}-"
-  apim_name_prefix = "${local.project}${local.domain}${var.environment.app_name}"
-
   apim = {
-    name           = "${local.apim_name_prefix}-apim-${var.environment.instance_number}"
-    autoscale_name = var.tier == "l" ? "${local.apim_name_prefix}-apim-autoscale-${var.environment.instance_number}" : null
+    name           = "${module.naming_convention.prefix}-apim-${module.naming_convention.suffix}"
+    autoscale_name = var.tier == "l" ? "${module.naming_convention.prefix}-apim-autoscale-${module.naming_convention.suffix}" : null
```
Comment (Krusty93) [inline infra/modules/azure_api_management/locals.tf:4]: `autoscale` should be abbreviated `as` (not part of this PR)
```suggestion
    autoscale_name = var.tier == "l" ? "${module.naming_convention.prefix}-apim-as-${module.naming_convention.suffix}" : null
```

File: infra/modules/azure_api_management/locals.tf [hunk 8]
```diff
@@ -1,18 +1,7 @@
...
-  location_short   = lookup(local.location_map, var.environment.location, "neu")
-  project          = "${var.environment.prefix}-${var.environment.env_short}-${local.location_short}"
-  domain           = var.environment.domain == null ? "-" : "-${var.environment.domain}-"
-  apim_name_prefix = "${local.project}${local.domain}${var.environment.app_name}"
-
   apim = {
-    name           = "${local.apim_name_prefix}-apim-${var.environment.instance_number}"
-    autoscale_name = var.tier == "l" ? "${local.apim_name_prefix}-apim-autoscale-${var.environment.instance_number}" : null
+    name           = "${module.naming_convention.prefix}-apim-${module.naming_convention.suffix}"
+    autoscale_name = var.tier == "l" ? "${module.naming_convention.prefix}-apim-as-${module.naming_convention.suffix}" : null
```
Comment (gunzip) [inline infra/modules/azure_api_management/locals.tf:4]: is this a breaking change? in case we need to bump this module to a major version

Comment (mamu0) [reply infra/modules/azure_api_management/locals.tf:4]: no don't worry, this module is new, it has never been used yet

**Review by Krusty93 — APPROVED** [review]

**Review by gunzip — APPROVED** [review]
